crypto/src/signature/secp256k1.rs

```diff
@@ -1,103 +1,101 @@
 use self::ecdsa_secp256k1::EcdsaSecp256k1Impl;
-use crate::{Algorithm, Error, KeyGenOption, PrivateKey, PublicKey};
+use crate::{Error, KeyGenOption};
 
 pub const PRIVATE_KEY_SIZE: usize = 32;
-pub const PUBLIC_KEY_SIZE: usize = 33;
-
-const ALGORITHM: Algorithm = Algorithm::Secp256k1;
 
 pub struct EcdsaSecp256k1Sha256;
 
+pub type PublicKey = k256::PublicKey;
+pub type PrivateKey = k256::SecretKey;
+
 impl EcdsaSecp256k1Sha256 {
-    pub fn keypair(option: Option<KeyGenOption>) -> Result<(PublicKey, PrivateKey), Error> {
+    pub fn keypair(option: Option<KeyGenOption>) -> (PublicKey, PrivateKey) {
         EcdsaSecp256k1Impl::keypair(option)
     }
-    pub fn sign(message: &[u8], sk: &PrivateKey) -> Result<Vec<u8>, Error> {
+
+    pub fn sign(message: &[u8], sk: &PrivateKey) -> Vec<u8> {
         EcdsaSecp256k1Impl::sign(message, sk)
     }
-    pub fn verify(message: &[u8], signature: &[u8], pk: &PublicKey) -> Result<bool, Error> {
+
+    pub fn verify(message: &[u8], signature: &[u8], pk: &PublicKey) -> Result<(), Error> {
         EcdsaSecp256k1Impl::verify(message, signature, pk)
     }
+
+    pub fn parse_public_key(payload: &[u8]) -> Result<PublicKey, Error> {
+        EcdsaSecp256k1Impl::parse_public_key(payload)
+    }
+
+    pub fn parse_private_key(payload: &[u8]) -> Result<PrivateKey, Error> {
+        EcdsaSecp256k1Impl::parse_private_key(payload)
+    }
 }
 
 mod ecdsa_secp256k1 {
-    use amcl::secp256k1::ecp;
     use arrayref::array_ref;
     use digest::Digest as _;
-    use iroha_primitives::const_vec::ConstVec;
     use rand::{rngs::OsRng, RngCore, SeedableRng};
     use rand_chacha::ChaChaRng;
     use signature::{Signer as _, Verifier as _};
     use zeroize::Zeroize;
 
-    use super::{ALGORITHM, PRIVATE_KEY_SIZE, PUBLIC_KEY_SIZE};
-    use crate::{Error, KeyGenOption, PrivateKey, PublicKey};
+    use super::{PrivateKey, PublicKey, PRIVATE_KEY_SIZE};
+    use crate::{Error, KeyGenOption};
 
     pub struct EcdsaSecp256k1Impl;
     type Digest = sha2::Sha256;
 
     impl EcdsaSecp256k1Impl {
-        pub fn public_key_compressed(pk: &PublicKey) -> Vec<u8> {
-            assert_eq!(pk.digest_function, ALGORITHM);
-            let mut compressed = [0u8; PUBLIC_KEY_SIZE];
-            ecp::ECP::frombytes(&pk.payload[..]).tobytes(&mut compressed, true);
-            compressed.to_vec()
-        }
-
-        pub fn keypair(option: Option<KeyGenOption>) -> Result<(PublicKey, PrivateKey), Error> {
-            let signing_key = match option {
-                Some(mut o) => match o {
+        pub fn keypair(option: Option<KeyGenOption>) -> (PublicKey, PrivateKey) {
+            let signing_key = option.map_or_else(
+                || PrivateKey::random(&mut OsRng),
+                |mut o| match o {
                     KeyGenOption::UseSeed(ref mut seed) => {
                         let mut s = [0u8; PRIVATE_KEY_SIZE];
```

Comment:
For `Ed25519Sha512` the seed is first hashed and then fed into `ChaChaRng` to generate the key. Here the order is reversed. Fishy stuff...

Reply:
Opened as a separate issue #4190